test/Prototypes/string_prototype/StringComparison.swift

```diff
@@ -633,3 +634,393 @@ where
   }
 }
 
+//
+// A "normalization segment" is a sequence that starts with a unicode scalar
+// value which has a normalization boundary before it, and includes all
+// subsequent unicode scalar values who do not.
+//
+// A "native index" is an index into the original String's code units. It can be
+// interchanged with many of the other String indices. TODO: I think this is
+// what UnicodeView solves, phrase it in those terms.
+//
+//
+//
+
+// Michael TODO: Why doesn't the following work?
+//
+// extension UnsignedInteger : _DefaultConstructible {
+//   init() { self = 0 }
+// }
+
+extension UInt16 : _DefaultConstructible {}
+
+typealias UTF16CodeUnitBuffer =
+  UnboundCapacity<_BoundedCapacity<_Array8<UInt16>>>
+
+// A normalization segment that is FCC-normalized. This is a collection of
+// normalized UTF16 code units.
+//
+// Note that indices into a normalized segment are not native indices, and do
+// not necessarily correspond to any particular code unit as they may of
+// undergone composition or decomposition. Thus, FCCNormalizedSegments are not
+// suitable for queries needing sub-segment granularity. However,
+// FCCNormalizedSegments are suitable for segment-level-or-coarser granularity
+// queries, which include any grapheme-level queries as segments are sub-
+// grapheme.
+//
+// TODO: Explore coalescing small segments together
+struct FCCNormalizedSegment : BidirectionalCollection {
+  let buffer: UTF16CodeUnitBuffer
+
+  init(_ buffer: UTF16CodeUnitBuffer) {
+    self.buffer = buffer
+  }
+  init() {
+    self.buffer = UTF16CodeUnitBuffer()
+  }
+
+  typealias Index = UTF16CodeUnitBuffer.Index
+
+  public var startIndex : Index {
+    return buffer.startIndex
+  }
+  public var endIndex : Index {
+    return buffer.endIndex
+  }
+  public subscript(i: Index) -> UInt16 {
+    return buffer[i]
+  }
+  public func index(after i: Index) -> Index {
+    return buffer.index(after: i)
+  }
+  public func index(before i: Index) -> Index {
+    return buffer.index(before: i)
+  }
+  public typealias SubSequence = BidirectionalSlice<FCCNormalizedSegment>
+}
+
+// Initialize an UnboundCapacity from any sequence
+extension UnboundCapacity {
```

Comment:
Don't we get an implementation of this automatically due to `UnboundCapacity`'s conformance to `RangeReplaceableCollection`?